ui/app/modifiers/code-mirror.ts

```diff
@@ -0,0 +1,59 @@
+import Modifier from 'ember-modifier';
+import codemirror from 'codemirror';
+
+import waypointHclMode from './utils/waypointHclMode';
+import 'codemirror/addon/edit/matchbrackets';
+import 'codemirror/addon/edit/closebrackets';
+import 'codemirror/addon/selection/active-line';
+import 'codemirror/addon/mode/simple';
+
+const _PRESET_DEFAULTS: codemirror.EditorConfiguration = {
+  theme: 'monokai',
+  lineNumbers: true,
+  cursorBlinkRate: 500,
+  matchBrackets: true,
+  autoCloseBrackets: true,
+  styleActiveLine: true,
+};
+interface Args {
+  positional: never;
+  named: {
+    value: string;
+    onInput: (value: string) => void;
+    options: Record<string, unknown>;
+  };
+}
+
+// call to create the CodeMirror waypoint HCL mode for syntax highlighting
+waypointHclMode();
+export default class CodeMirrorModifier extends Modifier<Args> {
+  _editor!: codemirror.Editor;
+
+  didInstall(): void {
+    this._setup();
+  }
+
```

Comment:
A thought for the future evolution of this component (no need to address in this PR)…

At some point we may want to implement the [`didUpdateArguments`](https://github.com/ember-modifier/ember-modifier#api) hook. This would mainly be so that if the calling code decides to send in a new `value` then the editor updates accordingly (i.e. `this._editor.setValue(newValue)`). Imagine, for example, a “reset” button.

It’d probably look something like this:

```ts
didUpdateArguments() {
  // this test might not be necessary if codemirror is smart about this sort of thing
  if (this.args.value !== this._editor.getValue()) {
    this._editor.setValue(this.args.value);
  }

  if (this.args.options) {
    for (let [key, value] of Object.entries(this.args.options)) {
      this._editor.setOption(key, value);
    }
  }
}
```